### ISSUE TITLE
PMIx dstore issue

### ISSUE THREAD

**Issue by hjelmn** [pmix/master#280]
I am running into a problem with the version of PMIx currently in master. When I run with a large number of processes I get errors from pmix_esh.c. Examples:

```
[nid20581:125718] PMIX ERROR: ERROR in file dstore/pmix_esh.c at line 2095
[nid20581:125718] PMIX ERROR: ERROR in file dstore/pmix_esh.c at line 2432
[nid20581:125718] PMIX ERROR: ERROR in file dstore/pmix_esh.c at line 1094
[nid20581:125718] PMIX ERROR: ERROR in file common/pmix_jobdata.c at line 97
[nid20581:125718] PMIX ERROR: ERROR in file common/pmix_jobdata.c at line 376
```

The errors go away if I up the default segment size. I do not see this bug with Open MPI 2.x.

**Comment by artpol84** [comment]
As per telecon we will add the env variable to enable runtime tuning of the segment size `PMIX_MCA_ESH_SEGMENT_SIZE`

**Comment by artpol84** [comment]
@rhc54, @hjelmn stated that this bug is not observed with Open MPI 2.x. I guess this means that 1.2 is not affected, isn't it?

**Comment by rhc54** [comment]
Not quite true. Remember, OMPI v2.x branch currently has an older version of PMIx in it - it does not have the current dstore code. OMPI master, however, does have the current head of the dstore code. So once we update OMPI v2.x, this same problem is likely to be seen there.

**Comment by artpol84** [comment]
Ok.

**Comment by artpol84** [comment]
I may have found the reason. @hjelmn will check today.

**Comment by artpol84** [comment]
In short the reason was that we are using special service markers called extension slots that are represented as a service key-value in the dstore.
We have a macro called `EXT_SLOT_SIZE` to calculate the size that is necessary for this key-value to reserve enough space when adding a regulare key-values. It turns out that this macro had a `key` field in it allowing arbitrary key name to be used to calculate kv size while the key name is actually fixed to  "EXTENSION_SLOT". And [there were places](https://github.com/pmix/master/blob/master/src/dstore/pmix_esh.c#L2129) in the code where `key` to be stored was used as an argument of  `EXT_SLOT_SIZE`. I think that if the key size was smaller than "EXTENSION_SLOT" not enough space was reserved which later caused a problem.

I guess that large number of ranks/keys was triggering it and increase of seg size was reducing number of mistakes hiding the error.

**Comment by rhc54** [comment]
Very kewl - thanks!

**Comment by artpol84** [comment]
We faced another problem that was reported [here](https://github.com/pmix/master/pull/282#issuecomment-277454198). Providing a copy here for the reference:
```
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2201
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2396
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 1094
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 97
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 376
[nid16980:96732] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/server/pmix_server.c at line 463
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2201
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 2396
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/dstore/pmix_esh.c at line 1094
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 97
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/common/pmix_jobdata.c at line 376
[nid16940:96735] PMIX ERROR: ERROR in file ../../../../../../../opal/mca/pmix/pmix2x/pmix/src/server/pmix_server.c at line 463
```

Here are 2 logs from the debugging session for the records:
```
[nid11605:05030] dstore/pmix_esh.c:2118:put_data_to_the_end: key pmix.nrank
[nid11605:05030] dstore/pmix_esh.c:2189:put_data_to_the_end: key pmix.nrank, rel start offset 4194201, rel end offset 4194236, abs shift 0 size 14
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack_int32 * 1
[nid11605:05030] pmix_bfrop_unpack: found 1 values for 1 provided storage
[nid11605:05030] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaab81eefe0, 1, 28 )
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack: 1 kvals
[nid11605:05030] pmix_bfrop_unpack_int32 * 1
[nid11605:05030] pmix_bfrop_unpack_byte * 12
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaab81e7ac8, 1, 14 )
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack_int32 * 1
[nid11605:05030] pmix: unpacked key pmix.nodeid
[nid11605:05030] dstore/pmix_esh.c:2203:pmix_sm_store: for rank 18533, replace flag 0
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_int32 * 1
[nid11605:05030] pmix_bfrop_pack_buffer( 0x2aaab8116c80, 0x2aaab81e7ac0, 1, 21 )
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_buffer( 0x2aaab8116c80, 0x2aaab81e7ac8, 1, 14 )
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_int32 * 1
[nid11605:05030] dstore/pmix_esh.c:2118:put_data_to_the_end: key pmix.nodeid
[nid11605:05030] dstore/pmix_esh.c:2189:put_data_to_the_end: key pmix.nodeid, rel start offset 4194236, rel end offset 4194273, abs shift 0 size 16
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] dstore/pmix_esh.c:2054:_get_data_region_by_offset
[nid11605:05030] dstore/pmix_esh.c:1968:set_rank_meta_info: nspace 3123445761, add rank 18533 offset 4194129 count 4 meta info
[nid11605:05030] dstore/pmix_esh.c:1025:_esh_store: for 3123445761:18534
[nid11605:05030] dstore/pmix_esh.c:1873:_get_track_elem_for_namespace: nspace 3123445761
[nid11605:05030] dstore/pmix_esh.c:2386:_store_data_for_rank: for rank 18534
[nid11605:05030] dstore/pmix_esh.c:1911:_get_rank_meta_info
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack_int32 * 1
[nid11605:05030] pmix_bfrop_unpack: found 1 values for 1 provided storage
[nid11605:05030] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaab81f8d60, 1, 28 )
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack: 1 kvals
[nid11605:05030] pmix_bfrop_unpack_int32 * 1
[nid11605:05030] pmix_bfrop_unpack_byte * 10
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaab81c6418, 1, 40 )
[nid11605:05030] pmix_bfrop_unpack_int16 * 1
[nid11605:05030] pmix_bfrop_unpack_int32 * 1
[nid11605:05030] pmix: unpacked key pmix.rank
[nid11605:05030] dstore/pmix_esh.c:2203:pmix_sm_store: for rank 18534, replace flag 0
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_int32 * 1
[nid11605:05030] pmix_bfrop_pack_buffer( 0x2aaab81dbe40, 0x2aaab81c6410, 1, 21 )
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_buffer( 0x2aaab81dbe40, 0x2aaab81c6418, 1, 40 )
[nid11605:05030] pmix_bfrop_pack_int16 * 1
[nid11605:05030] pmix_bfrop_pack_int32 * 1
[nid11605:05030] dstore/pmix_esh.c:2118:put_data_to_the_end: key pmix.rank
::::: global_offset = 0, offset = 0, id = 0, _data_segment_size = 4194304
```

```
[nid12418:273832] dstore/pmix_esh.c:2118:put_data_to_the_end: key pmix.nrank
[nid12418:273832] dstore/pmix_esh.c:2189:put_data_to_the_end: key pmix.nrank, rel start offset 4194201, rel end offset 4194236, abs shift 0 size 14
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack_int32 * 1
[nid12418:273832] pmix_bfrop_unpack: found 1 values for 1 provided storage
[nid12418:273832] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaad54d1d80, 1, 28 )
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack: 1 kvals
[nid12418:273832] pmix_bfrop_unpack_int32 * 1
[nid12418:273832] pmix_bfrop_unpack_byte * 12
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaab8188238, 1, 14 )
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack_int32 * 1
[nid12418:273832] pmix: unpacked key pmix.nodeid
[nid12418:273832] dstore/pmix_esh.c:2203:pmix_sm_store: for rank 18547, replace flag 0
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_int32 * 1
[nid12418:273832] pmix_bfrop_pack_buffer( 0x2aaab81d7f40, 0x2aaab8188230, 1, 21 )
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_buffer( 0x2aaab81d7f40, 0x2aaab8188238, 1, 14 )
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_int32 * 1
[nid12418:273832] dstore/pmix_esh.c:2118:put_data_to_the_end: key pmix.nodeid
[nid12418:273832] dstore/pmix_esh.c:2189:put_data_to_the_end: key pmix.nodeid, rel start offset 4194236, rel end offset 4194273, abs shift 0 size 16
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] dstore/pmix_esh.c:2054:_get_data_region_by_offset
[nid12418:273832] dstore/pmix_esh.c:1968:set_rank_meta_info: nspace 3123445761, add rank 18547 offset 4194129 count 4 meta info
[nid12418:273832] dstore/pmix_esh.c:1025:_esh_store: for 3123445761:18548
[nid12418:273832] dstore/pmix_esh.c:1873:_get_track_elem_for_namespace: nspace 3123445761
[nid12418:273832] dstore/pmix_esh.c:2386:_store_data_for_rank: for rank 18548
[nid12418:273832] dstore/pmix_esh.c:1911:_get_rank_meta_info
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack_int32 * 1
[nid12418:273832] pmix_bfrop_unpack: found 1 values for 1 provided storage
[nid12418:273832] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaad54d1d80, 1, 28 )
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack: 1 kvals
[nid12418:273832] pmix_bfrop_unpack_int32 * 1
[nid12418:273832] pmix_bfrop_unpack_byte * 10
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack_buffer( 0x2aaaafa75870, 0x2aaab81f3898, 1, 40 )
[nid12418:273832] pmix_bfrop_unpack_int16 * 1
[nid12418:273832] pmix_bfrop_unpack_int32 * 1
[nid12418:273832] pmix: unpacked key pmix.rank
[nid12418:273832] dstore/pmix_esh.c:2203:pmix_sm_store: for rank 18548, replace flag 0
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_int32 * 1
[nid12418:273832] pmix_bfrop_pack_buffer( 0x2aaab81c8340, 0x2aaab81f3890, 1, 21 )
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_buffer( 0x2aaab81c8340, 0x2aaab81f3898, 1, 40 )
[nid12418:273832] pmix_bfrop_pack_int16 * 1
[nid12418:273832] pmix_bfrop_pack_int32 * 1
[nid12418:273832] dstore/pmix_esh.c:2118:put_data_to_the_end: key pmix.rank
::::: global_offset = 0, offset = 0, id = 0, _data_segment_size = 4194304
```

**Comment by artpol84** [comment]
I will update with the precise analysis tomorrow.
The fix is in the PR #282

**Comment by artpol84** [comment]
I was able to reproduce locally:
```bash
$ ./pmix_test -n 4 --job-fence -c -t 30000
[artpol-VirtualBox:24866] PMIX ERROR: ERROR in file dstore/pmix_esh.c at line 2237
[artpol-VirtualBox:24866] PMIX ERROR: ERROR in file dstore/pmix_esh.c at line 2430
[artpol-VirtualBox:24866] PMIX ERROR: ERROR in file dstore/pmix_esh.c at line 1110
[artpol-VirtualBox:24866] PMIX ERROR: ERROR in file common/pmix_jobdata.c at line 97
[artpol-VirtualBox:24866] PMIX ERROR: ERROR in file common/pmix_jobdata.c at line 376
[artpol-VirtualBox:24866] PMIX ERROR: ERROR in file server/pmix_server.c at line 469
ERROR [cli_stages.c:148:cli_cleanup]: rank 0 with state 1 unexpectedly terminated.
ERROR [pmix_test.c:183:main]: Test was aborted!
```

Prerequisites:
* [My custom branch](https://github.com/artpol84/master/tree/test_suite/master) that has modified test suite which has and option of emulating large-scale job data (see #299 for more info).
* Patch below should be applied to it to emulate exact situation described in https://github.com/pmix/master/pull/282/commits/797e0a3fa99abee01785825fd5d8efde71321ca3

**Patch**:
```patch
diff --git a/src/dstore/pmix_esh.c b/src/dstore/pmix_esh.c
index 44edce1..340d036 100644
--- a/src/dstore/pmix_esh.c
+++ b/src/dstore/pmix_esh.c
@@ -2147,6 +2147,7 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
     global_offset = get_free_offset(dataseg);
     offset = global_offset % _data_segment_size;
 
+
     /* We should provide additional space at the end of segment to place EXTENSION_SLOT to have an ability to enlarge data for this rank.*/
     if (sizeof(size_t) + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE() > _data_segment_size) {
         /* this is an error case: segment is so small that cannot place evem a single key-value pair.
@@ -2156,6 +2157,19 @@ static size_t put_data_to_the_end(ns_track_elem_t *ns_info, seg_desc_t *dataseg,
                sizeof(size_t) + strlen(key) + 1 + sizeof(size_t) + size + EXT_SLOT_SIZE());
         return offset;
     }
+
+    int extsize = EXT_SLOT_SIZE();
+    int zerodata =  ESH_KEY_SIZE(key, 0);
+    int fulldata = ESH_KEY_SIZE(key, size);
+
+    if( (offset + fulldata + extsize > _data_segment_size) &&
+        (offset + zerodata + extsize < _data_segment_size) &&
+        (0 == strcmp(key,PMIX_HOSTNAME))  ){
+        /* key == PMIX_HOSTNAME is the last key in this rank's blob */
+        size = _data_segment_size - offset - zerodata - extsize;
+    }
+    
+
     if (offset + ESH_KEY_SIZE(key, size) + EXT_SLOT_SIZE() > _data_segment_size)  {
         id++;
         /* create a new data segment. */
diff --git a/test/utils.c b/test/utils.c
index 79d59da..9a2c03d 100644
--- a/test/utils.c
+++ b/test/utils.c
@@ -83,7 +83,8 @@ static void set_namespace(int nprocs, char *ranks, char *name, int nproc_tot)
     for (n=0; n < nproc_tot; n++) {
         pmix_info_t *kvp = &(info[8 + n]);
         pmix_info_t *info1;
-        char hname[256];
+#define HNAME_SIZE (2 * 1024)
+        char hname[HNAME_SIZE];
         int count = 5, k = 0;
         PMIX_INFO_CREATE(info1, count);
 
@@ -103,15 +104,12 @@ static void set_namespace(int nprocs, char *ranks, char *name, int nproc_tot)
         k++;
         INFO_KV_SET(&info1[k], PMIX_NODEID, uint32_t, n / nprocs);
         k++;
-        sprintf(hname,"cluster-node-%d",n / nprocs);
-        INFO_KV_SET(&info1[k], PMIX_HOSTNAME, string, hname);
-    }
-
-    {
-        int delay = 0;
-        while(delay ){
-            sleep(1);
+        int JJ;
+        for( JJ = 0; JJ < HNAME_SIZE; JJ++){
+            hname[JJ] = 'a';
         }
+        hname[HNAME_SIZE - 1] = '\0';
+        INFO_KV_SET(&info1[k], PMIX_HOSTNAME, string, hname);
     }
 
     volatile int in_progress = 1, rc;
```

**Comment by artpol84** [comment]
Conclusion: the problem was correctly identified and fixed

**Comment by jjhursey** [comment]
@artpol84 So #282 and #284 are ready to merge then, right?

**Comment by artpol84** [comment]
double-checking them now as we agreed on telecon

**Comment by jjhursey** [comment]
PRs #282 and #284 are merged in. I think this issue can be closed now. @hjelmn If you want PR #289 to be included in `v1.2` as part of this issue let us know.
